source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -157,8 +158,24 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
     if (slices_to_commit > 0) {
       read_buffer.commit(slices, slices_to_commit);
       if (callbacks_->shouldDrainReadBuffer()) {
-        callbacks_->setReadBufferReady();
-        keep_reading = false;
+        // Verify that SSL_get_read_ahead is disabled. SSL_pending does not provide an accurate
+        // answer when read-ahead is enabled. As far as we can tell BoringSSL does not implement
+        // read-ahead, so the odds of us ever failing this sanity check are effectively zero.
+        ASSERT(!SSL_get_read_ahead(rawSsl()));
+
+        // Query the SSL implementation for the number of bytes available for immediate read from
+        // internal buffers, and do one last read iteration if bytes are available. This is
+        // important to ensure read resumption works correctly after calls to
+        // Network::Connection::readDisable().
+        int pending_bytes = SSL_pending(rawSsl());
+        ASSERT(pending_bytes < 16 * 1024, "SSL record should be at most 16KB");
+        if (pending_bytes > 0) {
+          ASSERT(keep_reading);
+          reserve_size = pending_bytes;
```

Comment:
So I get the need to drain from the ssl buffer.  What I'm less sure of is if we explicitly set pending_bytes to the bytes already buffered, if we're guaranteed that SSL_read won't read more from transport, but will only consume the current record which is already buffered.  I was think of paging boringssl devs for that, but I think it'd be better to just regression test the behavior.

Is it possible to have a test where we send [lots of data][evenmoredata]
and verify that if the first read buffers [lots of data] and we tickle this corner case to drain it, that we won't also read and buffer evenmoredata?  Otherwise I'm concerned we'll trade lack of resumption for lack of ceding.  I think HighWatermarkReadResumptionProcessingHeaders probably lets you visually verify, but if SSL_read got greedy I don't think that test would fail (I think we'd need a unit test, but correct me if I'm wrong!)

Reply:
(NB: I don't know anything about how Envoy handles I/O.)

This is sufficient of an edge case on the abstraction that testing things seems worthwhile (for better or worse, OpenSSL's SSL API is supposed to look vaguely like a UNIX TCP socket). But `SSL_read` will currently never touch the transport if there's is already an unconsumed, decrypted record in the buffer. I wouldn't anticipate us changing that. Likewise, when readahead is off (currently always off and even if we did implement it, it'd be optional), non-zero SSL_pending implies there isn't also an undecrypted record in the buffer.

Though I think the intention of the API (especially if, like OpenSSL but unlike BoringSSL, you believe in renegotiation) is that callers not make as strong of assumptions about the correlation between SSL reads and transport reads. If there's a renegotiation (again, not applicable for BoringSSL), SSL_read may be blocked on the handshake, which may be blocked on transport *write* or some random callback. A more general implementation (closer to what Chromium does) might look something like:

RunReadStateMachine()
* If we're not ready to consume read from SSL (application layer doesn't want data, too much is buffered already, etc.), return. Ensure that RunReadStateMachine() is called again when this changes.
* Call SSL_read, independent of the state of the transport.
* If SSL_read returned data, great! Pass along to the application. Go back to the top of RunReadStateMachine() in case the application still wants data.
* If SSL_read failed with SSL_ERROR_WANT_READ, return. Ensure that RunReadStateMachine() is called again when the transport is readable.
* Ditto for other SSL_ERROR_WANT_* codes that we care to support. Though without renego or client 0-RTT, I think only WANT_READ matters here.

This doesn't depend on the state of the buffers, and it's always safe to call RunReadStateMachine(). However, one difference is RunReadStateMachine() will not pass data to an overbuffered application, even if the data is secretly available in SSL_read already. I think this is fine and conceptually simpler, but Antonio mentioned to me the intent was to pass that data along if available?

Reply:
alyssawilk: SslReadBufferLimitTest.DrainToSslRecordBoundary attempts to cover the case you describe by doing 10 10kb writes and verifying that the calls to doRead on the server connection with a configured high-watermark of 16KB deliver the data in 20KB chunks.

I reached out to davidben about to help confirm my understanding of the current implementation and how it's likely to evolve going forward since the fix implemented here ends up depending on some internal details how boringssl deals with partial reads.  Seems like the assumptions I'm making are fairly reasonable, but we may want to also think about implications they have on the transport socket contract and how likely it is for other transport socket implementations to fall on similar traps.

There is one other possible way to fix resumption in this case: explicitly schedule read resumption in cases where the read buffer is empty.
Change set: https://github.com/envoyproxy/envoy/compare/master...antoniovicente:ssl_read_resumption_alternate
Related comments: https://github.com/envoyproxy/envoy/issues/12304#issuecomment-664511408
This approach seems similar to the idea that davidben describes on how to ideally interact with SSL read/write.

A benefit of this alternate approach is that imposes less strict requirements on transport socket implementations but it does involve some possibly spurious synthetic fd readable events.  I have an old draft PR that optimizes away fd re-registrations on readDisable transitions which seems to fix this resumption issue, but its blocked on performance data and debugging of some test failures: https://github.com/envoyproxy/envoy/compare/master...antoniovicente:optimize_read_disable

Reply:
@yanavlasov So it would be better to do something similar to this alternate solution? https://github.com/envoyproxy/envoy/compare/master...antoniovicente:ssl_read_resumption_alternate

The issue with a state machine that specific to SSL handling is that it doesn't generalize well to other transport socket implementations.  I think the more eager event scheduling proposed in my alternate fix is the closest we can get to davidben's suggestions.

Reply:
I'm mildly inclined to avoid wake-ups every time.  If we don't think we can do a clean and cheap "is there buffered data to process" check I'd lean towards the original solution with extensive testing in case boring or openssl ever change.

I'd definitely like alternate opinions here on which of the ugly options are the least ugly.  cc @mattklein123 @snowp @ggreenway for thoughts

Reply:
I think this is related to the my question above. I'm confused why we need this in the first place. If the old impl set read to ready, why wouldn't the data be sucked out the next time read is ready?

Reply:
@mattklein123 In the normal case the current resumption logic works correctly.  The problem is when we factor in calls to readDisable(true) + readDisable(false) done as we trigger and untrigger the buffer high watermark

Calls to readDisable unregister and re-register the fd for events.  When fds are unregistered, any pending synthetic events on the fd are cleared, so the synthetic read event scheduled via setReadBufferReady is never delivered.  Resumption after readDisable(false) depends on either there being some bytes in the user-space read buffer or some bytes in the kernel's socket receive buffer.  When handling SSL traffic it is possible to have readable bytes in SSL internal buffers while there are 0 bytes in the stream read buffer and 0 bytes in the kernel buffer; when that happens fd re-registration does not lead to a wakeup despite the transport socket being able to produce bytes.

Reply:
> Calls to readDisable unregister and re-register the fd for events. When fds are unregistered, any pending synthetic events on the fd are cleared, so the synthetic read event scheduled via setReadBufferReady is never delivered. Resumption after readDisable(false) depends on either there being some bytes in the user-space read buffer or some bytes in the kernel's socket receive buffer. When handling SSL traffic it is possible to have readable bytes in SSL internal buffers while there are 0 bytes in the stream read buffer and 0 bytes in the kernel buffer; when that happens fd re-registration does not lead to a wakeup despite the transport socket being able to produce bytes.

Thanks this makes sense. At minimum, can you add more comments?

Another implementation idea: what if we had a transport socket specific way of asking "are there more bytes?" This could be overridden for TLS to look for kernel/user/within SSL. Then on readEnable() the check would pass and we would read again?

Reply:
See https://github.com/envoyproxy/envoy/pull/13772 instead. Your comment inspired me to think a bit more about cases where the resumption event is lost due to calls to readDisable. The connection impl can detect and remember when that happens and schedule resumption when needed.